Fix for turret tracking invisible pillbox

notsinned opened the pull request:
Fixes #20214 Tank turret re-aligns and no longer tracks camo pillbox when pillbox goes invisible after attacking.

notsinned reviewed (commented):
Fixed turret tracking units that go invisible after attacking. [#20214]

PunkPun commented:
Hi, thanks for you contribution.

It works as intended, but it seems that there's a better solution to this. Just replace the `target.Type == TargetType.Invalid` from the if clause above with `!target.IsValidFor(self)`. `IsValidFor` already checks for invalid target within the method.

notsinned commented:
Thank you for the feedback! I made the requested changes so hopefully it should be good now.

PunkPun reviewed (commented):
👍 

Please squash the commits

abcdefg30 commented on OpenRA.Mods.Common/Traits/Turreted.cs:
```diff
@@ -255,9 +255,10 @@ public bool FaceTarget(Actor self, in Target target)
+			if (!target.IsValidFor(self))
 			{
-				desiredDirection = WVec.Zero;
+				realignDesired = true;
+				MoveTurret();
```
This will probably break `RealignDelay`. Isn't setting `desiredDirection = WVec.Zero;` and letting `Tick` do the work enough?

PunkPun replied:
I've also had concerns, but it seems this is necessary for the unit to actually stop targeting. 
```C#
desiredDirection = WVec.Zero;
```
does nothing

PunkPun replied:
You are right, this is also not correct place for the fix

Changing `Target.Type != TargetType.Invalid` to `Target.IsValidFor(self)` everywhere in https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs#L126-L148 seems to fix the bug and also respect `RealignDelay`

I'd not advise changing this every else though till #17227 gets implemented / addressed

notsinned commented:
Sorry for the delay but the changes have been made and tested. Hopefully they're good now.

PunkPun commented on OpenRA.Mods.Common/Traits/Turreted.cs:
```diff
@@ -255,12 +255,6 @@ public bool FaceTarget(Actor self, in Target target)
-				desiredDirection = WVec.Zero;
-				return false;
-			}
-
```
This code should probably still stay

notsinned committed:
Fix for turret following invisible unit

Fixed turret tracking invisible units after being attacked

Fix for turret following invisible unit

Fix for turrent following invisible unit

PunkPun reviewed (approved):
Works perfectly. 

I tried to break it with other ways of target becoming invalid and different scenarios. Everything seems to be functioning as expected

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b9f14d08e683fc353d1936a2e1d8bad91e02d0e1...c1fd9e7a68fb00f77cc079ac69456a357ad9c271)
